### PR TITLE
Create a patch for retrying to list members in the join command

### PR DESCRIPTION
**Pull request by wongni**
*Issue #, if available:*

*Description of changes:*
When a new ETCD machine tries to join a cluster while the machine is not reachable the joining process fails at listing members without retrying. This PR adds the retry logic to this step so that the machine can join when it is reachable. 

Upstream PR: https://github.com/kubernetes-sigs/etcdadm/pull/316

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by jaxesn** [comment]
Can you open an usptream PR as well to etcdadm? cc @g-gaston

**Comment by wongni** [comment]
> Can you open an usptream PR as well to etcdadm? cc @g-gaston

I already opened one : https://github.com/kubernetes-sigs/etcdadm/pull/316

File: projects/kubernetes-sigs/etcdadm/CHECKSUMS [hunk 1]
```diff
@@ -1,2 +1,2 @@
-1a71735aea6e420988d308c847e1e7df1c146254160fd0ed1c99dcdfb3072a9e  _output/bin/etcdadm/linux-amd64/etcdadm
-37b087f5b30d1787f0f666553b563d6a8f3a02dd4f86e6b8a92a2ef77f665b9d  _output/bin/etcdadm/linux-arm64/etcdadm
+f34fe924adb93dc2e05b766d80adae2a821ad514fd77fca67aa516a1be96c22b  _output/bin/etcdadm/linux-amd64/etcdadm
+7e8429dee8352da69fa2b377f785c923b295da7b244f4e1905d6d1cba0350b04  _output/bin/etcdadm/linux-arm64/etcdadm
```
Comment (vivek-koppuru) [inline projects/kubernetes-sigs/etcdadm/CHECKSUMS:2]: Try removing this space here

File: projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch [hunk 2]
```diff
@@ -0,0 +1,90 @@
+From 35a9d8d692c24b89be947ba216980676670668fa Mon Sep 17 00:00:00 2001
+From: Wonkun Kim <wonkun@amazon.com>
+Date: Fri, 3 Jun 2022 15:33:17 -0500
+Subject: [PATCH] Update join.go
```
Comment (vivek-koppuru) [inline projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch:4]: Is it possible to change the commit message so it describes what it is, similar to the name of the patch file?

File: projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+--- a/cmd/join.go
++++ b/cmd/join.go
+@@ -132,6 +132,22 @@ func membership() phase {
+ 				return nil
+ 			}
+ 
++			var lastErr error
++			retrySteps := 1
++			if in.etcdAdmConfig.Retry {
++				retrySteps = constants.DefaultBackOffSteps
```
Comment (vivek-koppuru) [inline projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch:21]: Why does this have to be in a condition? Do we want the retrySteps to be 1 if this is not configured?

File: projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+-			ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultEtcdRequestTimeout)
+-			mresp, err := client.MemberList(ctx)
+-			cancel()
++
++			err = wait.ExponentialBackoff(opts, func() (bool, error) {
++				ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultEtcdRequestTimeout)
++				mresp, err := client.MemberList(ctx)
++				cancel()
++				if err != nil {
++					log.Warningf("[membership] Error listing members: %v, will retry after %s.", err, opts.Step())
```
Comment (vivek-koppuru) [inline projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch:50]: Is `[membership]` the right format for this warning message here?

File: projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch (hunk 3 above)
Comment (wongni) [reply projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch:21]: DefaultBackOffSteps is 5 and we want retrySteps to be one if Retry is false not to retry the step.

File: projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch (hunk 2 above)
Comment (wongni) [reply projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch]: Fixed.

File: projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch (hunk 4 above)
Comment (wongni) [reply projects/kubernetes-sigs/etcdadm/patches/0014-Add-retry-on-listing-members-in-join-command.patch:50]: Yes, it indicates that the function name is `membership`.

**Pull request merged**